### ISSUE TITLE
improve translate-c for libcurl simple example

### ISSUE THREAD

**Issue by cup** [ziglang/zig#988]
For example, with C you can do this:

https://github.com/curl/curl/blob/master/docs/examples/simple.c

and other languages:

http://rosettacode.org/wiki/HTTP

**Comment by andrewrk** [comment]
You can use libcurl in zig too.

I wonder how far this will get you:

`zig translate-c simple.c`

**Comment by svnpenn** [comment]
not very far :(

~~~
$ zig translate-c \
--libc-include-dir 'C:\cygwin64\usr\x86_64-w64-mingw32\sys-root\mingw\include' \
simple.c
unexpected error from clang: too many errors emitted, stopping now
D:\Desktop\zig-0.2.0\lib\zig\include\mmintrin.h:81:40: error: passing '__v2si'
(aka 'int') to parameter of incompatible type
'__attribute__((__vector_size__(2 * sizeof(int)))) int' (vector of 2 'int' values)
    return __builtin_ia32_vec_ext_v2si((__v2si)__m, 0);
                                       ^
~~~

**Comment by andrewrk** [comment]
this issue is a duplicate of #910 

but I'm going to repurpose it to be an issue to improve translate-c of this curl example.

**Comment by svnpenn2** [comment]
@andrewrk I am posting with alt - it seems @cup has been blocked from this repo?

**Comment by andrewrk** [comment]
Hmm that does appear to be the case. However I did not leave any notes for myself on why, so it may have been a mistake. My apologies. @cup is unblocked now.

I am hiding both this comment and the one above as offtopic so that people can view Zig's issues without meta discussion such as this.

**Comment by cup** [comment]
https://github.com/ziglang/zig/issues/2007

**Comment by cup4** [comment]
@andrewrk it seems @cup has been blocked again - will you please advise?

**Comment by daurnimator** [comment]
The curl sample file now fails with:
```
/home/daurnimator/src/zig/988.zig:2393:36: error: expected type 'enum_unnamed_33', found 'comptime_int'
        _ = curl_easy_setopt(curl, @as(CURLoption, CURLOPT_URL), "https://example.com");
                                   ^
/home/daurnimator/src/zig/988.zig:1414:25: note: enum_unnamed_33 declared here
const enum_unnamed_33 = extern enum {
                        ^
/home/daurnimator/src/zig/lib/std/start.zig:258:17: error: expected return type of main to be 'void', '!void', 'noreturn', 'u8', or '!u8'
                @compileError(bad_main_ret);
                ^
```

**Comment by andrewrk** [comment]
The example above now translates to:

```zig
pub export fn main() c_int {
    var curl: ?*CURL = undefined;
    var res: CURLcode = undefined;
    curl = curl_easy_init();
    if (curl != null) {
        _ = curl_easy_setopt(curl, @intToEnum(CURLoption, CURLOPT_URL), "https://example.com");
        _ = curl_easy_setopt(curl, @intToEnum(CURLoption, CURLOPT_FOLLOWLOCATION), @as(c_long, 1));
        res = curl_easy_perform(curl);
        if (@enumToInt(res) != @bitCast(c_uint, CURLE_OK)) _ = fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
        curl_easy_cleanup(curl);
    }
    return 0;
}
```

It would be nice if Zig could figure out to improve the enum handling in this way instead:

```zig
pub export fn main() c_int {
    var curl: ?*CURL = undefined;
    var res: CURLcode = undefined;
    curl = curl_easy_init();
    if (curl != null) {
        _ = curl_easy_setopt(curl, .CURLOPT_URL, "https://example.com");
        _ = curl_easy_setopt(curl, .CURLOPT_FOLLOWLOCATION, @as(c_long, 1));
        res = curl_easy_perform(curl);
        if (res != .CURLE_OK) _ = fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
        curl_easy_cleanup(curl);
    }
    return 0;
}
```

But I think it's safe to say that this issue is solved.

**Comment by lithdew** [comment]
Just wanted to comment here for anyone trying to use libcurl that enum handling already works in the way @andrewrk mentioned.

```zig
const std = @import("std");
const c = @cImport(@cInclude("curl/curl.h"));

fn wrap(result: anytype) !void {
    switch (@enumToInt(result)) {
        c.CURLE_OK => return,
        else => unreachable,
    }
}

pub fn main() !void {
    const curl = c.curl_easy_init();
    if (curl == null) return error.InitFailed;
    defer c.curl_easy_cleanup(curl);

    try wrap(c.curl_easy_setopt(curl, .CURLOPT_URL, "https://client.tlsfingerprint.io:8443/"));
    try wrap(c.curl_easy_setopt(curl, .CURLOPT_FOLLOWLOCATION, @as(c_long, 1)));

    if (c.curl_easy_perform(curl) == .CURLE_OK) {
        std.debug.print("It works!\n", .{});
    }
}
```
